src/elements/name_section.rs

```diff
@@ -10,92 +10,156 @@ const NAME_TYPE_LOCAL: u8 = 2;
-	},
+pub struct NameSection {
+	/// Module name subsection.
+	module_name_subsection: Option<ModuleNameSubsection>,
+
+	/// Function name subsection.
+	function_name_subsection: Option<FunctionNameSubsection>,
+
+	/// Local name subsection.
+	local_name_subsection: Option<LocalNameSubsection>,
+}
+
+impl NameSection {
+	/// Creates a new name section.
+	pub fn new(module_name_subsection: Option<ModuleNameSubsection>,
+			   function_name_subsection: Option<FunctionNameSubsection>,
+			   local_name_subsection: Option<LocalNameSubsection>) -> Self {
+		Self {
+			module_name_subsection,
+			function_name_subsection,
+			local_name_subsection,
+		}
+	}
+
+	/// Module name subsection of this section.
+	pub fn module_name_subsection(&self) -> Option<&ModuleNameSubsection> {
+		self.module_name_subsection.as_ref()
+	}
+
+	/// Module name subsection of this section (mutable).
+	pub fn module_name_subsection_mut(&mut self) -> &mut Option<ModuleNameSubsection> {
+		&mut self.module_name_subsection
+	}
+
+	/// Functions name subsection of this section.
+	pub fn function_name_subsection(&self) -> Option<&FunctionNameSubsection> {
+		self.function_name_subsection.as_ref()
+	}
+
+	/// Functions name subsection of this section (mutable).
+	pub fn function_name_subsection_mut(&mut self) -> &mut Option<FunctionNameSubsection> {
+		&mut self.function_name_subsection
+	}
+
+	/// Local name subsection of this section.
+	pub fn local_name_subsection(&self) -> Option<&LocalNameSubsection> {
+		self.local_name_subsection.as_ref()
+	}
+
+	/// Local name subsection of this section (mutable).
+	pub fn local_name_subsection_mut(&mut self) -> &mut Option<LocalNameSubsection> {
+		&mut self.local_name_subsection
+	}
 }
 
 impl NameSection {
 	/// Deserialize a name section.
 	pub fn deserialize<R: io::Read>(
 		module: &Module,
 		rdr: &mut R,
-	) -> Result<NameSection, Error> {
-		let name_type: u8 = VarUint7::deserialize(rdr)?.into();
-		let name_payload_len: u32 = VarUint32::deserialize(rdr)?.into();
-		let name_section = match name_type {
-			NAME_TYPE_MODULE => NameSection::Module(ModuleNameSection::deserialize(rdr)?),
-			NAME_TYPE_FUNCTION => NameSection::Function(FunctionNameSection::deserialize(module, rdr)?),
-			NAME_TYPE_LOCAL => NameSection::Local(LocalNameSection::deserialize(module, rdr)?),
-			_ => {
-				let mut name_payload = vec![0u8; name_payload_len as usize];
-				rdr.read(&mut name_payload)?;
-				NameSection::Unparsed {
-					name_type,
-					name_payload,
-				}
-			}
-		};
-		Ok(name_section)
+	) -> Result<Self, Error> {
+		let mut module_name_subsection: Option<ModuleNameSubsection> = None;
+		let mut function_name_subsection: Option<FunctionNameSubsection> = None;
+		let mut local_name_subsection: Option<LocalNameSubsection> = None;
+
+		loop {
+			let subsection_type: u8 = match VarUint7::deserialize(rdr) {
+				Ok(raw_subsection_type) => raw_subsection_type.into(),
+				// todo: be more selective detecting no more subsection
```

Comment:
can you explain?

Reply:
This approach of parsing is very similar to [section deserialization](https://github.com/paritytech/parity-wasm/blob/master/src/elements/section.rs#L84) (and this cmt comes directly from it:)). 

But it seems a little bit unreliable because there could be different types of IO errors, and we are interested only in `UnexpectedEof` here. I don't dig too much to parity-wasm internals, but in my point of view, the corner case of this problem is [here](https://github.com/paritytech/parity-wasm/blob/master/src/elements/mod.rs#L227) when all IO errors map to Error::HeapOther, so it is impossible in section and subsections parsing to distinguish `UnexpectedEof` from other errors.